### PR TITLE
Add support for multi-topic claim and limits

### PR DESCRIPTION
**Pull request by basharal**
This change allows Marshal Consumer to handle multiple topics and
honor MaximumClaims across all topics.

### CODE REVIEW

File: marshal/consumer.go [hunk 1]
```diff
@@ -49,15 +49,14 @@ type ConsumerOptions struct {
 	StrictOrdering bool
 
 	// The maximum number of claims this Consumer is allowed to hold simultaneously.
-	// This limits the number of partitions claimed. Set to 0 (default) to allow
-	// an unlimited number of claims.
+	// This limits the number of partitions even in the case of multiple topics.
+	// Set to 0 (default) to allow an unlimited number of claims.ß
```
Comment (zorkian) [inline marshal/consumer.go]: ß?

File: marshal/consumer.go [hunk 2]
```diff
@@ -77,49 +76,61 @@ type Consumer struct {
...
+func (m *Marshaler) NewConsumer(topicNames []string, options ConsumerOptions) (*Consumer, error) {
 	if m.Terminated() {
 		return nil, errors.New("Marshaler has terminated, no new consumers can be created")
 	}
 
-	// Ensure the user hasn't set a claim limit and put us into topic mode
-	if options.ClaimEntireTopic && options.MaximumClaims != 0 {
-		return nil, errors.New("ClaimEntireTopic and MaximumClaims are incompatible options")
+	if len(topicNames) > 1 && !options.ClaimEntireTopic {
+		return nil, errors.New("ClaimEntireTopic must be set for more than one topic")
```
Comment (zorkian) [inline marshal/consumer.go]: Is this a code restriction somehow?

In theory it should be possible to claim arbitrary partitions from any number of given topics.

File: marshal/consumer.go [hunk 3]
```diff
@@ -77,49 +76,61 @@ type Consumer struct {
...
+	} else if len(topicNames) == 0 {
+		return nil, errors.New("length of topicNames must be postive")
```
Comment (zorkian) [inline marshal/consumer.go]: How about `must provide at least one topic`?

File: marshal/consumer.go (hunk 2 above)
Comment (basharal) [inline marshal/consumer.go]: Yes. It is code restriction...claimpartitions() vs. claimtopics()...claim partitions assumes there is a single topic...i can make it claim from topics arbitrarily if you want...i thought it's closer to current behavior

File: marshal/consumer.go (hunk 1 above)
Comment (basharal) [inline marshal/consumer.go]: oops...done

File: marshal/consumer.go (hunk 3 above)
Comment (basharal) [inline marshal/consumer.go]: done...i thought it was easier to grep :)

**Comment by zorkian** [comment]
Merged manually due to conflicts: https://github.com/dropbox/marshal/commit/717ae1fdb55f06aa9510ba12e5ab3a71cd674311

**Pull request closed without merging**